crates/re_viewer/src/store_hub.rs

```diff
@@ -1,14 +1,174 @@
+use ahash::HashMap;
+use itertools::Itertools;
+use re_arrow_store::{DataStoreConfig, DataStoreStats};
 use re_data_store::StoreDb;
-use re_log_types::{StoreId, StoreKind};
+use re_log_types::{ApplicationId, StoreId, StoreKind};
+use re_viewer_context::StoreContext;
 
-/// Stores many [`StoreDb`]s of recordings and blueprints.
 #[derive(Default)]
 pub struct StoreHub {
+    selected_rec_id: Option<StoreId>,
+    application_id: Option<ApplicationId>,
+    blueprint_by_app_id: HashMap<ApplicationId, StoreId>,
+    store_dbs: StoreBundle,
+}
+
+#[derive(Default)]
+/// Convenient information used for `MemoryPanel`
+pub struct StoreHubStats {
+    pub blueprint_stats: DataStoreStats,
+    pub blueprint_config: DataStoreConfig,
+    pub recording_stats: DataStoreStats,
+    pub recording_config: DataStoreConfig,
+}
+
+/// Interface for accessing all blueprints and recordings
+///
+/// The [`StoreHub`] provides access to the [`StoreDb`] instances that are used
+/// to store both blueprints and recordings.
+///
+/// Internally, the [`StoreHub`] tracks which [`ApplicationId`] and `recording
+/// id` ([`StoreId`]) are currently selected in the viewer. These can be configured
+/// using [`StoreHub::set_recording_id`] and [`StoreHub::set_app_id`] respectively.
+impl StoreHub {
+    /// Add a [`StoreBundle`] to the [`StoreHub`]
+    pub fn add_bundle(&mut self, bundle: StoreBundle) {
+        self.store_dbs.append(bundle);
+    }
+
+    /// Get a read-only [`StoreContext`] from the [`StoreHub`] if one is available.
+    ///
+    /// All of the returned references to blueprints and recordings will have a
+    /// matching [`ApplicationId`].
+    pub fn read_context(&mut self) -> Option<StoreContext<'_>> {
+        // If we have an app-id, then use it to look up the blueprint or else
+        // create the default blueprint.
+        let blueprint_id = self.application_id.as_ref().map(|app_id| {
+            self.blueprint_by_app_id
+                .entry(app_id.clone())
+                .or_insert_with(|| StoreId::from_string(StoreKind::Blueprint, app_id.clone().0))
+        });
+
+        // TODO(jleibs) entry is what I want, but holds a mutable reference. We know that
+        // unwrap won't fail since we just created the entry, but
```

Comment:
but..?

Reply:
oops... "but it would be nice to do this without using unwrap()"